# Khám phá dữ liệu

Notebook này để khám phá dữ liệu.

## Import những thư viện cần thiết

In [1]:
import numpy as np
import pandas as pd

## Đọc dữ liệu từ file

File dữ liệu của đồ án sau bước thu thập dữ liệu đang nằm tại địa chỉ '../thu_thap_du_lieu/animal_crawler/animal_info.csv'. Ta sẽ tiến hành đọc dữ liệu đã thu thập và lưu vào DataFrame `animal_df`

In [2]:
animal_df = pd.read_csv('../thu_thap_du_lieu/animal_crawler/animals_info.csv')

Hiển thị 5 dòng đầu tiên trong dữ liệu để xem thử đã đọc đúng hay chưa

In [3]:
pd.reset_option('display.max_colwidth')
animal_df.head(5)

Name   Kingdom    Phylum   Subphylum     Class         Order  \
0     Grey Wolf  Animalia  Chordata  Vertebrata  Mammalia     Carnivora   
1         Tiger  Animalia  Chordata  Vertebrata  Mammalia     Carnivora   
2    Brown Bear  Animalia  Chordata  Vertebrata  Mammalia     Carnivora   
3    Blue Whale  Animalia  Chordata  Vertebrata  Mammalia  Artiodactyla   
4  Killer Whale  Animalia  Chordata  Vertebrata  Mammalia  Artiodactyla   

     Suborder           Family         Genus                Species  ...  \
0  Caniformia          Canidae         Canis            Canis lupus  ...   
1  Feliformia          Felidae      Panthera        Panthera tigris  ...   
2  Caniformia          Ursidae         Ursus           Ursus arctos  ...   
3         NaN  Balaenopteridae  Balaenoptera  Balaenoptera musculus  ...   
4         NaN      Delphinidae       Orcinus           Orcinus orca  ...   

  Top speed      Weight     Height      Length  \
0   75 km/h    16-60 kg   80-85 cm  105-160 cm   
1   96 km/h   65-306 kg        NaN  200-390 cm   
2   56 km/h  100-635 kg  70-153 cm   1.4-2.8 m   
3   20 km/h   100-160 t        NaN     25-30 m   
4   45 km/h       3-6 t        NaN       6-9 m   

                                          Attributes  \
0  Nocturnal,Carnivore,Scavenger,Terrestrial,Curs...   
1  Nocturnal,Carnivore,Cursorial,Terrestrial,Ambu...   
2  Crepuscular,Omnivore,Terrestrial,Altricial,Nom...   
3  Carnivore,Piscivores,Aquatic,Precocial,Natator...   
4  Diurnal,Carnivore,Aquatic,Precocial,Cosmopolit...   

                                        Distribution  \
0  {'Geography': {'Continents': 'Asia, Europe, No...   
1  {'Geography': {'Continents': 'Asia', 'Subconti...   
2  {'Geography': {'Continents': 'Asia, Europe, No...   
3  {'Geography': {'Continents': 'Africa, South Am...   
4  {'Geography': {'Continents': 'Africa, Oceania,...   

                                              Habits                   Diet  \
0  {'Group name': 'pack, route, rout', 'Lifestyle...   Carnivore,,Scavenger   
1  {'Lifestyle': 'Cursorial, Terrestrial, Ambush ...              Carnivore   
2  {'Group name': 'sloth, sleuth', 'Lifestyle': '...               Omnivore   
3  {'Group name': 'pod, gam, herd, school, mod', ...  Carnivore,,Piscivores   
4  {'Group name': 'pod, gam, herd, school, mod', ...              Carnivore   

                                       Mating_Habits  \
0  {'Mating behavior': 'Monogamy', 'Reproduction ...   
1  {'Mating behavior': 'Polygyny', 'Reproduction ...   
2  {'Mating behavior': 'Polygynandry,', 'Reproduc...   
3  {'Mating behavior': 'Monogamy', 'Reproduction ...   
4  {'Mating behavior': 'Polygynandry', 'Reproduct...   

                                          Population  
0  {'Population trend': 'Stable', 'Population sta...  
1  {'Population trend': 'Decreasing', 'Population...  
2  {'Population trend': 'Stable', 'Population sta...  
3  {'Population trend': 'Increasing', 'Population...  
4       {'Population status': 'Data deficient (DD)'}  

[5 rows x 22 columns]

Xem xét dữ liệu trên thì thấy một số dòng bị lỗi, cần loại bỏ.
Cụ thể ở đây là thu thập dữ liệu về thú nhưng có một số dòng có giá trị cột 'Kingdom' không phải là 'Animalia'. Ở đây ta sẽ xóa luôn những dòng như thế, vì dữ liệu khá lớn nên xóa bớt một vài dòng có lẽ sẽ không có ảnh hưởng nhiều.


In [4]:
animal_df = animal_df[animal_df['Kingdom'] == 'Animalia']

## Dữ liệu có bao nhiêu dòng, bao nhiêu cột?

In [5]:
animal_df.shape

(28525, 22)

## Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng cho chúng ta biết các thông tin về một con vật, tất cả các dòng đều có cùng ý nghĩa như nhau.

## Dữ liệu có các dòng bị lặp không?

Chúng ta sử dụng hàm 'duplicated' trong pandas để kiểm tra dòng đó có phải bị lặp trong DataFrame hay không. Nếu kết quả là dòng đó bị lặp sẽ trả về giá trị 1, ngược lại sẽ là 0. Kế tiếp chúng ta sử dùng hàm 'sum' để tính tất cả giá trị trả về của hàm 'duplicated'. Nếu kết quả > 0 thì chứng tỏ có tồn tại dòng bị lặp. 

In [6]:
animal_df.duplicated().sum()

0

Vậy là dữ liệu của chúng ta không tồn lại dòng nào bị lặp lại.

## Mỗi cột có ý nghĩa gì?

Sau khi thu thập dữ liệu xong, nhóm em đã tự phân tích ý nghĩa của mỗi trong file dữ liệu vừa thu thập, sau đó viết vào file `col_meaning.xlsx`. Đọc file excel này để biết được ý nghĩa của mỗi cột.

In [7]:
xlsx = pd.ExcelFile('col_meaning.xlsx')
col_meaning_df =pd.read_excel(xlsx,"Sheet1")
pd.set_option('display.max_colwidth', 0)
col_meaning_df


Column  \
0   Name              
1   Kingdom           
2   Phylum            
3   Subphylum         
4   Class             
5   Order             
6   Suborder          
7   Family            
8   Genus             
9   Species           
10  Population size   
11  Life span         
12  Top speed         
13  Weight            
14  Height            
15  Length            
16  Attributes        
17  Distribution      
18  Habits            
19  Diet              
20  Mating_Habits     
21  Population        

                                                                                                                                                  Description  
0   Name of Animal                                                                                                                                             
1   Giới là đơn vị phân loại lớn nhất bao gồm các ngành sinh vật có chung những đặc điểm nhất định.                                                            
2   Ngành: Một đơn vị phân loại ở cấp dưới giới và trên lớp.                                                                                                   
3   Phân Ngành: là một bậc phân loại nằm trung gian giữa ngành và lớp hoặc phân thứ ngành hoặc liên lớp                                                        
4   Lớp: Một mức độ phân loại động vật trong một ngành. Các lớp sau đó được chia nhỏ thành các nhóm khác được gọi là bộ                                        
5   Bộ: là một cấp nằm giữa lớp và họ.                                                                                                                         
6   Phân bộ: một nhóm thực vật hoặc động vật có liên quan, có quan hệ gần gũi hơn so với một bộ nhưng ít giống nhau hơn so với một họ.                         
7   Họ: Bộ được chia thành các họ, họ sẽ được chia thành các chi                                                                                               
8   Chi: còn gọi là giống là một đơn vị phân loại sinh học dùng để chỉ một hoặc một nhóm loài có kiểu hình tương tự và mối quan hệ tiến hóa gần gũi với nhau.  
9   Loài: là một nhóm các cá thể sinh vật có những đặc điểm sinh học tương đối giống nhau và có khả năng giao phối với nhau và sinh sản ra thế hệ tương lai.   
10  Ước tính số cá thể đang sống của loài                                                                                                                      
11  Tuổi thọ                                                                                                                                                   
12  Tốc độ                                                                                                                                                     
13  Cân nặng                                                                                                                                                   
14  Chiều cao                                                                                                                                                  
15  Chiều dài                                                                                                                                                  
16  Một số thông tin chính về loài vật                                                                                                                         
17  Phân bổ: Nơi sinh sống                                                                                                                                     
18  Thói quen                                                                                                                                                  
19  Chế độ ăn (ăn cỏ, ăn tạp,…)                                                                                                                                
20  Thói quen giao phối                                                                                                    

## Các cột lúc này đang có kiểu dữ liệu gì?

In [8]:
animal_df.dtypes

Name               object
Kingdom            object
Phylum             object
Subphylum          object
Class              object
Order              object
Suborder           object
Family             object
Genus              object
Species            object
Population size    object
Life span          object
Top speed          object
Weight             object
Height             object
Length             object
Attributes         object
Distribution       object
Habits             object
Diet               object
Mating_Habits      object
Population         object
dtype: object

## Tính xem tình trạng của mỗi cột dữ liệu hiện tại như thế nào?

### Với mỗi cột ta sẽ tính:
- Tỉ lệ mất mát: missing_ratio
- Số lượng giá trị khác nhau: num_values
- Số lần xuất hiện của từng giá trị: value

Viết hàm "col_info" để tính và lưu kết quả vào mảng "cat_col_info_df"

In [9]:
def col_info(col, array):
    temp = pd.DataFrame(array[col])
    temp = temp.dropna()
    temp = np.array(temp[col])
    data = []
    # missing_ratio
    data.append(round((100 - (len(temp) / len(array[col]) * 100)), 1))
    # num_values
    temp = np.unique(temp, return_counts=True)
    data.append(len(temp[0]))
    # value
    temp = pd.Series(data = temp[1], index = temp[0])
    data.append(temp)
    # num_values
    return data

In [10]:
cat_col_info_df = pd.DataFrame(index = ['missing_ratio', 'num_values', 'value'])
for i in animal_df.columns:
    if animal_df[i].dtype not in [float, int]:
        cat_col_info_df[i] = col_info(i, animal_df)
        
pd.reset_option('display.max_colwidth')
cat_col_info_df

Name  \
missing_ratio                                                0.0   
num_values                                                 28525   
value          ? Nycticebus linglom      1
AEECL's sportive l...   

                                      Kingdom  \
missing_ratio                             0.0   
num_values                                  1   
value          Animalia    28525
dtype: int64   

                                                         Phylum  \
missing_ratio                                               0.0   
num_values                                                    2   
value          Chordata    23659
Mollusca     4866
dtype: int64   

                                      Subphylum  \
missing_ratio                              64.4   
num_values                                    1   
value          Vertebrata    10152
dtype: int64   

                                                           Class  \
missing_ratio                                                0.0   
num_values                                                    10   
value          Amphibia           3393
Aves              1011...   

                                                           Order  \
missing_ratio                                                1.5   
num_values                                                   118   
value          Accipitriformes    255
Adapedonta          10
...   

                                                        Suborder  \
missing_ratio                                               91.0   
num_values                                                    13   
value          Alcedini            109
Caniformia          19...   

                                                          Family  \
missing_ratio                                                0.0   
num_values                                                   840   
value          Abrocomidae           10
Abronicidae          ...   

                                                           Genus  \
missing_ratio                                                0.0   
num_values                                                  6127   
value          Aaadonta        6
Abditomys       1
Abeillia  ...   

                                                         Species  ...  \
missing_ratio                                                0.0  ...   
num_values                                                 28523  ...   
value          "Calamita" melanorabdotus          1
"Calamita...  ...   

                                                       Top speed  \
missing_ratio                                               98.0   
num_values                                                   124   
value          0.24 km/h    1
0.26 km/h    1
0.27 km/h    1
0...   

                                                          Weight  \
missing_ratio                                               90.0   
num_values                                                  2232   
value          0-0 kg            3
0.010-0.796 g     1
0.02-2...   

                                                          Height  \
missing_ratio                                               98.9   
num_values                                                   245   
value          0.5 m          1
0.7-1.3 m      1
0.75-1 m    ...   

                                                          Length  \
missing_ratio                                               89.7   
num_values                                                  2101   
value          0.5 m          2
0.5-1.8 m      1
0.6-2.43 m  ...   

                                                      Attributes  \
missing_ratio                                                0.0   
num_values                                                  3078   
value          Algivore,Detritivore,Herbivore,Fossorial,Polyg...   

                                                    Distribution  \
missing_ratio            

## Xử lý tách các cột có kiểu dictionary thành nhiều cột

Đầu tiên các cột dictionary đều có kiểu là string nên ta phải xử lý chuyển các cột trên thành dict

In [11]:
for i in animal_df:
    if animal_df[i][0].find('{') != -1:
        animal_df[i] = animal_df[i].apply(eval)

Viết hàm "split_col" để split các cột có kiểu dữ liệu là dict thành nhiều cột

In [52]:
def split_col(df):
    df_split_animals_info = pd.DataFrame()
    for col in df.columns:
        if type(df[col][0]) == dict:
            df_split_animals_info = pd.concat([df_split_animals_info, df[col].apply(pd.Series, dtype = object)], axis = 1)
        else:
            df_col = pd.DataFrame(df[col])
            df_split_animals_info = pd.concat([df_split_animals_info, df_col], axis = 1)
    return df_split_animals_info
    

In [47]:
df_split_animals_info = pd.DataFrame()
a= pd.DataFrame(animal_df['Name'])
df_split_animals_info = pd.concat([df_split_animals_info, a], axis = 1)
df_split_animals_info = pd.concat([df_split_animals_info, a], axis = 1)
df_split_animals_info

Name                         Name
0                        Grey Wolf                    Grey Wolf
1                            Tiger                        Tiger
2                       Brown Bear                   Brown Bear
3                       Blue Whale                   Blue Whale
4                     Killer Whale                 Killer Whale
...                            ...                          ...
29350            Córdoba cinclodes            Córdoba cinclodes
29351          Western hemispingus          Western hemispingus
29352  Sulphur-bearded reedhaunter  Sulphur-bearded reedhaunter
29355        McConnell's spinetail        McConnell's spinetail
29356              Junín canastero              Junín canastero

[28525 rows x 2 columns]

In [48]:
df_split_animals_info = split_col(animal_df)
pd.set_option('display.max_columns', None)
df_split_animals_info

Name   Kingdom    Phylum   Subphylum     Class  \
0                        Grey Wolf  Animalia  Chordata  Vertebrata  Mammalia   
1                            Tiger  Animalia  Chordata  Vertebrata  Mammalia   
2                       Brown Bear  Animalia  Chordata  Vertebrata  Mammalia   
3                       Blue Whale  Animalia  Chordata  Vertebrata  Mammalia   
4                     Killer Whale  Animalia  Chordata  Vertebrata  Mammalia   
...                            ...       ...       ...         ...       ...   
29350            Córdoba cinclodes  Animalia  Chordata         NaN      Aves   
29351          Western hemispingus  Animalia  Chordata         NaN      Aves   
29352  Sulphur-bearded reedhaunter  Animalia  Chordata         NaN      Aves   
29355        McConnell's spinetail  Animalia  Chordata         NaN      Aves   
29356              Junín canastero  Animalia  Chordata         NaN      Aves   

               Order    Suborder           Family         Genus  \
0          Carnivora  Caniformia          Canidae         Canis   
1          Carnivora  Feliformia          Felidae      Panthera   
2          Carnivora  Caniformia          Ursidae         Ursus   
3       Artiodactyla         NaN  Balaenopteridae  Balaenoptera   
4       Artiodactyla         NaN      Delphinidae       Orcinus   
...              ...         ...              ...           ...   
29350  Passeriformes         NaN      Furnariidae     Cinclodes   
29351  Passeriformes         NaN       Thraupidae    Sphenopsis   
29352  Passeriformes         NaN      Furnariidae   Cranioleuca   
29355  Passeriformes         NaN      Furnariidae    Synallaxis   
29356  Passeriformes         NaN      Furnariidae      Asthenes   

                         Species Population size   Life span Top speed  \
0                    Canis lupus         400,000   10-20 yrs   75 km/h   
1                Panthera tigris     2,154-3,159   10-15 yrs   96 km/h   
2                   Ursus arctos         200,000   20-50 yrs   56 km/h   
3          Balaenoptera musculus      10-25 Thou   80-90 yrs   20 km/h   
4                   Orcinus orca          50,000  30-100 yrs   45 km/h   
...                          ...             ...         ...       ...   
29350    Cinclodes comechingonus             NaN         NaN       NaN   
29351        Sphenopsis ochracea             NaN         NaN       NaN   
29352  Limnoctites sulphuriferus             NaN         NaN       NaN   
29355     Synallaxis macconnelli             NaN         NaN       NaN   
29356           Asthenes virgata             NaN         NaN       NaN   

           Weight     Height      Length  \
0        16-60 kg   80-85 cm  105-160 cm   
1       65-306 kg        NaN  200-390 cm   
2      100-635 kg  70-153 cm   1.4-2.8 m   
3       100-160 t        NaN     25-30 m   
4           3-6 t        NaN       6-9 m   
...           ...        ...         ...   
29350         NaN        NaN         NaN   
29351         NaN        NaN         NaN   
29352         NaN        NaN         NaN   
29355         NaN        NaN         NaN   
29356         NaN        NaN         NaN   

                                              Attributes  \
0      Nocturnal,Carnivore,Scavenger,Terrestrial,Curs...   
1      Nocturnal,Carnivore,Cursorial,Terrestrial,Ambu...   
2      Crepuscular,Omnivore,Terrestrial,Altricial,Nom...   
3      Carnivore,Piscivores,Aquatic,Precocial,Natator...   
4      Diurnal,Carnivore,Aquatic,Precocial,Cosmopolit...   
...                                                  ...   
29350                  Terrestrial,Migrating,starts with   
29351              Terrestrial,Not a migrant,starts with   
29352                  Terrestrial,Migrating,starts with   
29355              Terrestrial,Not a migrant,starts with   
29356              Terrestrial,Not a migrant,starts with   

                                               Geography  \
0      {'Continents': 'Asia, Europe, North America, A...   
1    

Sau khi split dữ liệu có số dòng là 28525 và số cột là 37, tăng lên 15 cột

Nhưng nhìn lại dữ liệu vẫn còn cột là dictionary như "Geograpy", như vậy ta phải split thêm 1 lần nữa

In [53]:
df_split_animals_info = split_col(df_split_animals_info)
df_split_animals_info.head(100)

Name   Kingdom    Phylum   Subphylum     Class  \
0           Grey Wolf  Animalia  Chordata  Vertebrata  Mammalia   
1               Tiger  Animalia  Chordata  Vertebrata  Mammalia   
2          Brown Bear  Animalia  Chordata  Vertebrata  Mammalia   
3          Blue Whale  Animalia  Chordata  Vertebrata  Mammalia   
4        Killer Whale  Animalia  Chordata  Vertebrata  Mammalia   
..                ...       ...       ...         ...       ...   
106  Black Rhinoceros  Animalia  Chordata  Vertebrata  Mammalia   
107           Markhor  Animalia  Chordata  Vertebrata  Mammalia   
109     Eurasian Lynx  Animalia  Chordata  Vertebrata  Mammalia   
110   Dromedary Camel  Animalia  Chordata  Vertebrata  Mammalia   
114    Common Warthog  Animalia  Chordata  Vertebrata  Mammalia   

              Order    Suborder           Family         Genus  \
0         Carnivora  Caniformia          Canidae         Canis   
1         Carnivora  Feliformia          Felidae      Panthera   
2         Carnivora  Caniformia          Ursidae         Ursus   
3      Artiodactyla         NaN  Balaenopteridae  Balaenoptera   
4      Artiodactyla         NaN      Delphinidae       Orcinus   
..              ...         ...              ...           ...   
106  Perissodactyla         NaN   Rhinocerotidae       Diceros   
107    Artiodactyla         NaN          Bovidae         Capra   
109       Carnivora  Feliformia          Felidae          Lynx   
110    Artiodactyla         NaN        Camelidae       Camelus   
114    Artiodactyla         NaN           Suidae  Phacochoerus   

                    Species Population size   Life span   Top speed  \
0               Canis lupus         400,000   10-20 yrs     75 km/h   
1           Panthera tigris     2,154-3,159   10-15 yrs     96 km/h   
2              Ursus arctos         200,000   20-50 yrs     56 km/h   
3     Balaenoptera musculus      10-25 Thou   80-90 yrs     20 km/h   
4              Orcinus orca          50,000  30-100 yrs     45 km/h   
..                      ...             ...         ...         ...   
106        Diceros bicornis     4,880-5,000   30-49 yrs  42-58 km/h   
107         Capra falconeri           5,754   10-13 yrs     17 km/h   
109               Lynx lynx      70-80 Thou   13-25 yrs     80 km/h   
110     Camelus dromedarius             NaN   40-50 yrs     64 km/h   
114  Phacochoerus africanus         Unknown   12-18 yrs         NaN   

           Weight      Height      Length  \
0        16-60 kg    80-85 cm  105-160 cm   
1       65-306 kg         NaN  200-390 cm   
2      100-635 kg   70-153 cm   1.4-2.8 m   
3       100-160 t         NaN     25-30 m   
4           3-6 t         NaN       6-9 m   
..            ...         ...         ...   
106  800-1,400 kg  140-180 cm     3-3.8 m   
107     32-110 kg   65-115 cm  132-186 cm   
109       8-30 kg    60-75 cm   80-130 cm   
110    300-600 kg     1.7-2 m       2-3 m   
114     45-150 kg  63.5-85 cm   0.9-1.5 m   

                                            Attributes  \
0    Nocturnal,Carnivore,Scavenger,Terrestrial,Curs...   
1    Nocturnal,Carnivore,Cursorial,Terrestrial,Ambu...   
2    Crepuscular,Omnivore,Terrestrial,Altricial,Nom...   
3    Carnivore,Piscivores,Aquatic,Precocial,Natator...   
4    Diurnal,Carnivore,Aquatic,Precocial,Cosmopolit...   
..                                                 ...   
106  Diurnal,Herbivore,Folivore,Lignivore,Terrestri...   
107  Diurnal,Herbivore,Folivore,Graminivore,Terrest...   
109  Crepuscular,Nocturnal,Carnivore,Terrestrial,Te...   
110  Diurnal,Herbivore,Folivore,Graminivore,Terrest...   
114  Diurnal,Omnivore,Scavenger,Terrestrial,Altrici...   

                                            Continents  \
0                  Asia, Europe, North America, Africa   
1                                                 Asia   
2                          Asia, Europe, North America   
3    Africa, South America, Oceania, North America,...   
4    Africa, Oceania, Antarctica, Nort

Sau khi split dữ liệu tăng thêm 9 cột

Lúc này ta tính lại missing_ratio xem các cột sau khi split như thế nào nhé

In [54]:
cat_col_info_df = pd.DataFrame(index = ['missing_ratio', 'num_values', 'value'])
for i in df_split_animals_info.columns:
    if df_split_animals_info[i].dtype not in [float, int]:
        cat_col_info_df[i] = col_info(i, df_split_animals_info)

cat_col_info_df

Name  \
missing_ratio                                                0.0   
num_values                                                 28525   
value          ? Nycticebus linglom      1
AEECL's sportive l...   

                                      Kingdom  \
missing_ratio                             0.0   
num_values                                  1   
value          Animalia    28525
dtype: int64   

                                                         Phylum  \
missing_ratio                                               0.0   
num_values                                                    2   
value          Chordata    23659
Mollusca     4866
dtype: int64   

                                      Subphylum  \
missing_ratio                              64.4   
num_values                                    1   
value          Vertebrata    10152
dtype: int64   

                                                           Class  \
missing_ratio                                                0.0   
num_values                                                    10   
value          Amphibia           3393
Aves              1011...   

                                                           Order  \
missing_ratio                                                1.5   
num_values                                                   118   
value          Accipitriformes    255
Adapedonta          10
...   

                                                        Suborder  \
missing_ratio                                               91.0   
num_values                                                    13   
value          Alcedini            109
Caniformia          19...   

                                                          Family  \
missing_ratio                                                0.0   
num_values                                                   840   
value          Abrocomidae           10
Abronicidae          ...   

                                                           Genus  \
missing_ratio                                                0.0   
num_values                                                  6127   
value          Aaadonta        6
Abditomys       1
Abeillia  ...   

                                                         Species  \
missing_ratio                                                0.0   
num_values                                                 28523   
value          "Calamita" melanorabdotus          1
"Calamita...   

                                                 Population size  \
missing_ratio                                               93.7   
num_values                                                   638   
value          0.45-1 M        1
1 Mln           2
1,000     ...   

                                                       Life span  \
missing_ratio                                               93.6   
num_values                                                   581   
value          1 yr            2
1-10 yrs        2
1-12.5 yrs...   

                                                       Top speed  \
missing_ratio                                               98.0   
num_values                                                   124   
value          0.24 km/h    1
0.26 km/h    1
0.27 km/h    1
0...   

                                                          Weight  \
missing_ratio                                               90.0   
num_values                                                  2232   
value          0-0 kg            3
0.010-0.796 g     1
0.02-2...   

                                                          Height  \
missing_ratio                                               98.9   
num_values                                                   245   
value          0.5 m          1
0.7-1.3 m      1
0.75-1 m    ...   

                                                          Length  \
missing_ratio                                

## Tiếp tục tiền xử lý dữ liệu nhé

### Loại bỏ một số cột không cần thiết

Sau khi tách các cột thành dictionary, đã xuất hiện một số cột không cần thiết mà tỷ lệ miss lại khá cao, ta cần vứt bỏ những cột này cho dễ xử lý về sau.

In [55]:
cols = ['Oceans', 'Introduced countries', 'Seas', 'Islands', 'Predators', "Bird's call"]
for col in cols:
    df_split_animals_info  = df_split_animals_info.drop(columns  = col)
df_split_animals_info

Name   Kingdom    Phylum   Subphylum     Class  \
0                        Grey Wolf  Animalia  Chordata  Vertebrata  Mammalia   
1                            Tiger  Animalia  Chordata  Vertebrata  Mammalia   
2                       Brown Bear  Animalia  Chordata  Vertebrata  Mammalia   
3                       Blue Whale  Animalia  Chordata  Vertebrata  Mammalia   
4                     Killer Whale  Animalia  Chordata  Vertebrata  Mammalia   
...                            ...       ...       ...         ...       ...   
29350            Córdoba cinclodes  Animalia  Chordata         NaN      Aves   
29351          Western hemispingus  Animalia  Chordata         NaN      Aves   
29352  Sulphur-bearded reedhaunter  Animalia  Chordata         NaN      Aves   
29355        McConnell's spinetail  Animalia  Chordata         NaN      Aves   
29356              Junín canastero  Animalia  Chordata         NaN      Aves   

               Order    Suborder           Family         Genus  \
0          Carnivora  Caniformia          Canidae         Canis   
1          Carnivora  Feliformia          Felidae      Panthera   
2          Carnivora  Caniformia          Ursidae         Ursus   
3       Artiodactyla         NaN  Balaenopteridae  Balaenoptera   
4       Artiodactyla         NaN      Delphinidae       Orcinus   
...              ...         ...              ...           ...   
29350  Passeriformes         NaN      Furnariidae     Cinclodes   
29351  Passeriformes         NaN       Thraupidae    Sphenopsis   
29352  Passeriformes         NaN      Furnariidae   Cranioleuca   
29355  Passeriformes         NaN      Furnariidae    Synallaxis   
29356  Passeriformes         NaN      Furnariidae      Asthenes   

                         Species Population size   Life span Top speed  \
0                    Canis lupus         400,000   10-20 yrs   75 km/h   
1                Panthera tigris     2,154-3,159   10-15 yrs   96 km/h   
2                   Ursus arctos         200,000   20-50 yrs   56 km/h   
3          Balaenoptera musculus      10-25 Thou   80-90 yrs   20 km/h   
4                   Orcinus orca          50,000  30-100 yrs   45 km/h   
...                          ...             ...         ...       ...   
29350    Cinclodes comechingonus             NaN         NaN       NaN   
29351        Sphenopsis ochracea             NaN         NaN       NaN   
29352  Limnoctites sulphuriferus             NaN         NaN       NaN   
29355     Synallaxis macconnelli             NaN         NaN       NaN   
29356           Asthenes virgata             NaN         NaN       NaN   

           Weight     Height      Length  \
0        16-60 kg   80-85 cm  105-160 cm   
1       65-306 kg        NaN  200-390 cm   
2      100-635 kg  70-153 cm   1.4-2.8 m   
3       100-160 t        NaN     25-30 m   
4           3-6 t        NaN       6-9 m   
...           ...        ...         ...   
29350         NaN        NaN         NaN   
29351         NaN        NaN         NaN   
29352         NaN        NaN         NaN   
29355         NaN        NaN         NaN   
29356         NaN        NaN         NaN   

                                              Attributes  \
0      Nocturnal,Carnivore,Scavenger,Terrestrial,Curs...   
1      Nocturnal,Carnivore,Cursorial,Terrestrial,Ambu...   
2      Crepuscular,Omnivore,Terrestrial,Altricial,Nom...   
3      Carnivore,Piscivores,Aquatic,Precocial,Natator...   
4      Diurnal,Carnivore,Aquatic,Precocial,Cosmopolit...   
...                                                  ...   
29350                  Terrestrial,Migrating,starts with   
29351              Terrestrial,Not a migrant,starts with   
29352                  Terrestrial,Migrating,starts with   
29355              Terrestrial,Not a migrant,starts with   
29356              Terrestrial,Not a migrant,starts with   

                                              Continents  \
0                    Asia, Europe, North America, Africa   
1    

Tại đây ta sẽ tách các cột có thuộc tính cách nhau bằng dấu phẩy thành 1 list

Luu vào "element" là tên các cột có các thuộc tính phân tách bằng dấu phẩy

Các cột sau khi xử lý sẽ được lưu vào biến "splits"

In [56]:
element = np.array(['Attributes', 'Continents', 'Subcontinents', 'Countries', 'Regions', 'Wwf biomes', 'Group name', 'Diet', 'Biogeographical realms', 'Lifestyle',  'Seasonal behavior'])

splits = [df_split_animals_info[col].str.split(pat = ',| and |  ', expand = False) for col in element]
splits = pd.DataFrame(splits).T
# print(splits.shape)
# print(splits.head())


Sau khi tách xong ta tiến hành ghép 2 dataframe lại thành "data_df"

In [57]:
df_split_animals_info.drop(element, axis=1, inplace=True)
data_df = pd.concat([df_split_animals_info, splits], axis = 1)
# data_df = data_df.drop(columns = data_df.columns[0], axis = 1)
data_df
# df_split_animals_info

Name   Kingdom    Phylum   Subphylum     Class  \
0                        Grey Wolf  Animalia  Chordata  Vertebrata  Mammalia   
1                            Tiger  Animalia  Chordata  Vertebrata  Mammalia   
2                       Brown Bear  Animalia  Chordata  Vertebrata  Mammalia   
3                       Blue Whale  Animalia  Chordata  Vertebrata  Mammalia   
4                     Killer Whale  Animalia  Chordata  Vertebrata  Mammalia   
...                            ...       ...       ...         ...       ...   
29350            Córdoba cinclodes  Animalia  Chordata         NaN      Aves   
29351          Western hemispingus  Animalia  Chordata         NaN      Aves   
29352  Sulphur-bearded reedhaunter  Animalia  Chordata         NaN      Aves   
29355        McConnell's spinetail  Animalia  Chordata         NaN      Aves   
29356              Junín canastero  Animalia  Chordata         NaN      Aves   

               Order    Suborder           Family         Genus  \
0          Carnivora  Caniformia          Canidae         Canis   
1          Carnivora  Feliformia          Felidae      Panthera   
2          Carnivora  Caniformia          Ursidae         Ursus   
3       Artiodactyla         NaN  Balaenopteridae  Balaenoptera   
4       Artiodactyla         NaN      Delphinidae       Orcinus   
...              ...         ...              ...           ...   
29350  Passeriformes         NaN      Furnariidae     Cinclodes   
29351  Passeriformes         NaN       Thraupidae    Sphenopsis   
29352  Passeriformes         NaN      Furnariidae   Cranioleuca   
29355  Passeriformes         NaN      Furnariidae    Synallaxis   
29356  Passeriformes         NaN      Furnariidae      Asthenes   

                         Species Population size   Life span Top speed  \
0                    Canis lupus         400,000   10-20 yrs   75 km/h   
1                Panthera tigris     2,154-3,159   10-15 yrs   96 km/h   
2                   Ursus arctos         200,000   20-50 yrs   56 km/h   
3          Balaenoptera musculus      10-25 Thou   80-90 yrs   20 km/h   
4                   Orcinus orca          50,000  30-100 yrs   45 km/h   
...                          ...             ...         ...       ...   
29350    Cinclodes comechingonus             NaN         NaN       NaN   
29351        Sphenopsis ochracea             NaN         NaN       NaN   
29352  Limnoctites sulphuriferus             NaN         NaN       NaN   
29355     Synallaxis macconnelli             NaN         NaN       NaN   
29356           Asthenes virgata             NaN         NaN       NaN   

           Weight     Height      Length  \
0        16-60 kg   80-85 cm  105-160 cm   
1       65-306 kg        NaN  200-390 cm   
2      100-635 kg  70-153 cm   1.4-2.8 m   
3       100-160 t        NaN     25-30 m   
4           3-6 t        NaN       6-9 m   
...           ...        ...         ...   
29350         NaN        NaN         NaN   
29351         NaN        NaN         NaN   
29352         NaN        NaN         NaN   
29355         NaN        NaN         NaN   
29356         NaN        NaN         NaN   

                                                   Biome  \
0      [Wetlands , Shrubland, Forest, Mountains, Anth...   
1      [Forest, Rainforest, Montane forest, Riparian,...   
2                                     [Forest, Woodland]   
3           [Oceanic pelagic zone, Marine, Neritic zone]   
4      [Oceanic pelagic zone, Marine, Neritic zone, E...   
...                                                  ...   
29350                          [Grassland, Rocky areas ]   
29351                                           [Forest]   
29352                 [Freshwater, Shrubland, Wetlands ]   
29355                      [Anthropogenic biome, Forest]   
29356                             [Grassland, Shrubland]   

                                  Climate Mating behavior  \
0      [Temperate, Cold, Polar, Tropical]        Monogamy   
1  

Xử lý tại các cột list, những giá trị nan sẽ được thay thế thành một list rỗng []

In [58]:
data_df[element] = data_df[element].fillna(data_df.notna().applymap(lambda x: x or []))
# data_df = data_df.drop(columns = data_df.columns[0], axis = 1)
data_df

Name   Kingdom    Phylum   Subphylum     Class  \
0                        Grey Wolf  Animalia  Chordata  Vertebrata  Mammalia   
1                            Tiger  Animalia  Chordata  Vertebrata  Mammalia   
2                       Brown Bear  Animalia  Chordata  Vertebrata  Mammalia   
3                       Blue Whale  Animalia  Chordata  Vertebrata  Mammalia   
4                     Killer Whale  Animalia  Chordata  Vertebrata  Mammalia   
...                            ...       ...       ...         ...       ...   
29350            Córdoba cinclodes  Animalia  Chordata         NaN      Aves   
29351          Western hemispingus  Animalia  Chordata         NaN      Aves   
29352  Sulphur-bearded reedhaunter  Animalia  Chordata         NaN      Aves   
29355        McConnell's spinetail  Animalia  Chordata         NaN      Aves   
29356              Junín canastero  Animalia  Chordata         NaN      Aves   

               Order    Suborder           Family         Genus  \
0          Carnivora  Caniformia          Canidae         Canis   
1          Carnivora  Feliformia          Felidae      Panthera   
2          Carnivora  Caniformia          Ursidae         Ursus   
3       Artiodactyla         NaN  Balaenopteridae  Balaenoptera   
4       Artiodactyla         NaN      Delphinidae       Orcinus   
...              ...         ...              ...           ...   
29350  Passeriformes         NaN      Furnariidae     Cinclodes   
29351  Passeriformes         NaN       Thraupidae    Sphenopsis   
29352  Passeriformes         NaN      Furnariidae   Cranioleuca   
29355  Passeriformes         NaN      Furnariidae    Synallaxis   
29356  Passeriformes         NaN      Furnariidae      Asthenes   

                         Species Population size   Life span Top speed  \
0                    Canis lupus         400,000   10-20 yrs   75 km/h   
1                Panthera tigris     2,154-3,159   10-15 yrs   96 km/h   
2                   Ursus arctos         200,000   20-50 yrs   56 km/h   
3          Balaenoptera musculus      10-25 Thou   80-90 yrs   20 km/h   
4                   Orcinus orca          50,000  30-100 yrs   45 km/h   
...                          ...             ...         ...       ...   
29350    Cinclodes comechingonus             NaN         NaN       NaN   
29351        Sphenopsis ochracea             NaN         NaN       NaN   
29352  Limnoctites sulphuriferus             NaN         NaN       NaN   
29355     Synallaxis macconnelli             NaN         NaN       NaN   
29356           Asthenes virgata             NaN         NaN       NaN   

           Weight     Height      Length  \
0        16-60 kg   80-85 cm  105-160 cm   
1       65-306 kg        NaN  200-390 cm   
2      100-635 kg  70-153 cm   1.4-2.8 m   
3       100-160 t        NaN     25-30 m   
4           3-6 t        NaN       6-9 m   
...           ...        ...         ...   
29350         NaN        NaN         NaN   
29351         NaN        NaN         NaN   
29352         NaN        NaN         NaN   
29355         NaN        NaN         NaN   
29356         NaN        NaN         NaN   

                                                   Biome  \
0      [Wetlands , Shrubland, Forest, Mountains, Anth...   
1      [Forest, Rainforest, Montane forest, Riparian,...   
2                                     [Forest, Woodland]   
3           [Oceanic pelagic zone, Marine, Neritic zone]   
4      [Oceanic pelagic zone, Marine, Neritic zone, E...   
...                                                  ...   
29350                          [Grassland, Rocky areas ]   
29351                                           [Forest]   
29352                 [Freshwater, Shrubland, Wetlands ]   
29355                      [Anthropogenic biome, Forest]   
29356                             [Grassland, Shrubland]   

                                  Climate Mating behavior  \
0      [Temperate, Cold, Polar, Tropical]        Monogamy   
1  

### Chuyển cột Top speed sang kiểu số

In [59]:
def to_float(element):
    try:
        return float(element)
    except ValueError:
        return np.nan

In [60]:
def trans_speed(speed):
    if pd.isna(speed): return np.nan
    speed = [speed[:-4].strip(), speed[-4:].strip()]
    if speed[1] == 'km/h':
        x = 1
    elif speed[1] == 'mph':
        x = 1.61
    sp = speed[0].split('-')
    if len(sp) == 1:
        return float(sp[0]) * x
    else:
        return (float(sp[0]) + float(sp[1])) / 2 * x


In [61]:
data_df['Top speed'] = data_df['Top speed'].apply(trans_speed)


### Chuyển cột life_span sang float

In [62]:
def trans_life_span(life):
    if pd.isna(life): return np.nan
    life = str(life)
    life = life.upper()
    life = life.replace('UPPER', '')
    life = life.replace('ABOVE', '')
    life = life.replace('OVER','')
    life = life.replace('mos', ' MONTHS ')
    life = life.replace('yr', ' YEARS ')
    life = life.replace('YRS', ' YEARS ')
    life = life.replace('YEARS', ' YEARS ')
    life = life.split('-')
    low = to_float(life[0])
    high = to_float(life[-1].split()[0].strip())
    x = life[-1].split()[-1].strip()
    if np.isnan(low):
        low = high
    t = 1.0
    if x == 'MONTHS':
        t = 1 / 12
    elif x == 'YEARS':
        t = 1.0
    low = low * t
    high = high * t
    return (low, high)

In [63]:
data_df['Life span'] = data_df['Life span'].apply(trans_life_span)
#data_df['Life span'].dropna()

## Thống kê các chỉ số cho các cột có kiểu dữ liệu là list

In [64]:
def stat_list_col(col, array):
    temp = pd.DataFrame(array[col])
    temp = temp.dropna()
    temp = np.array(temp[col])
    data = []
    # missing_ratio
    data.append(round((100 - (len(temp) / len(array[col]) * 100)), 1))
    # num_values
    temp = np.hstack(temp)
    temp = list(map(str.strip, temp)) # Xử lý các dấu cách đầu và cuối mỗi chuỗi
    temp = np.unique(temp, return_counts=True)
    data.append(len(temp[0]))
    # value
    temp = pd.Series(data = temp[1], index = temp[0])
    data.append(temp)
    # num_values
    return data

#     total = data_df[col].tolist()
#     total = np.hstack(total)
#     value, times = np.unique(total, return_counts = True)
    

In [65]:
list_col_info_df = pd.DataFrame(index = ['missing_ratio', 'num_values', 'value'])
for i in data_df:
    if type(data_df[i][0]) == list:
        list_col_info_df[i] = stat_list_col(i, data_df)
    pd.set_option('display.max_columns', None)

list_col_info_df

Biome  \
missing_ratio                                                0.0   
num_values                                                    43   
value          Abyssal zone                4
Agricultural    ...   

                                                         Climate  \
missing_ratio                                                0.0   
num_values                                                     6   
value          Arid          299
Cold          305
Libya     ...   

                                                      Attributes  \
missing_ratio                                                0.0   
num_values                                                   139   
value          Aestivation               71
Algivore         ...   

                                                      Continents  \
missing_ratio                                                0.0   
num_values                                                   132   
value                            11908
Africa             191...   

                                                   Subcontinents  \
missing_ratio                                                0.0   
num_values                                                    80   
value                            40
Afghanistan        1
Afro...   

                                                       Countries  \
missing_ratio                                                0.0   
num_values                                                   258   
value                            6611
Afghanistan         95
...   

                                                         Regions  \
missing_ratio                                                0.0   
num_values                                                   423   
value                            1120
Acre                 6
...   

                                                      Wwf biomes  \
missing_ratio                                                0.0   
num_values                                                   266   
value                                        179
Afghanistan ...   

                                                      Group name  \
missing_ratio                                                0.0   
num_values                                                   400   
value                            1
"bind"            1
"littl...   

                                                            Diet  \
missing_ratio                                                0.0   
num_values                                                    30   
value                            2972
Algivore            18
...   

                                          Biogeographical realms  \
missing_ratio                                                0.0   
num_values                                                   419   
value                                               600
Afgha...   

                                                       Lifestyle  \
missing_ratio                                                0.0   
num_values                                                    47   
value                                 9659
Aestivation       ...   

                                               Seasonal behavior  
missing_ratio                                                0.0  
num_values                                                     8  
value                                                        ...

### Lưu dữ liệu đã tiền xử lý vào file 'animals_split_info.csv' để tiếp tục sử dụng trong các phần sau

In [71]:
data_df.to_csv('animals_split_info.csv')